iointel/src/agent_methods/tools_before_rebase/duckduckgo.py

```diff
@@ -0,0 +1,114 @@
+import asyncio
+from typing import Any, Dict, List, Optional
+
+from pydantic import BaseModel, ConfigDict, model_validator
+
+
+class DuckDuckGoSearchAPIWrapper(BaseModel):
+    """Wrapper for DuckDuckGo Search API.
+
+    Free and does not require any setup.
+    """
+
+    region: Optional[str] = "wt-wt"
+    """
+    See https://pypi.org/project/duckduckgo-search/#regions
+    """
+    safesearch: str = "moderate"
+    """
+    Options: strict, moderate, off
+    """
+    time: Optional[str] = "y"
+    """
+    Options: d, w, m, y
+    """
+    max_results: int = 5
+    backend: str = "auto"
+    """
+    Options: auto, html, lite
+    """
+
+    model_config = ConfigDict(
+        extra="forbid",
+    )
+
+    @model_validator(mode="before")
+    @classmethod
+    def validate_environment(cls, values: Dict) -> Any:
+        """Validate that python package exists in environment."""
+        try:
+            from duckduckgo_search import DDGS  # noqa: F401
+        except ImportError:
+            raise ImportError(
+                "Could not import duckduckgo-search python package. "
+                "Please install it with `pip install -U duckduckgo-search`."
+            )
+        return values
+
+    def _ddgs(
+        self, query: str, max_results: Optional[int] = None
+    ) -> List[Dict[str, str]]:
+        """Run query through DuckDuckGo text search and return results."""
+        from duckduckgo_search import DDGS
+
+        with DDGS() as ddgs:
+            ddgs_gen = ddgs.text(
+                query,
+                region=self.region,  # type: ignore[arg-type]
+                safesearch=self.safesearch,
+                timelimit=self.time,
+                max_results=max_results or self.max_results,
+                backend=self.backend,
+            )
+            if ddgs_gen:
+                return [r for r in ddgs_gen]
+        return []
+
+    def run(self, query: str) -> str:
+        """Run query through DuckDuckGo and return concatenated results."""
+        results = self._ddgs_text(query)
```

Comment:
```suggestion
        results = self.results(query)
```
as there is no `_ddgs_text` method in the class

Reply:
Just gonna delete this method, i think self.results is enough